pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -3405,6 +3487,445 @@ def SeismicActivityExceptionRule(model, k, t):
+    )
+
+    if model.config.hydraulics == Hydraulics.post_process:
+        """
+        For the post process method, the pressure is computed using the
+        Hazen-Williams equation in a separate stand alone method "_hazen_williams_head". In this method,
+        the hydraulics block is solved alone for the objective of minimizing total cost of pumps.
+        """
+        mh.p_effective_Pipeline_diameter = Param(
+            model.s_LLA,
+            default=0,
+            initialize=0,
+            mutable=True,
+            units=model.model_units["diameter"],
+            doc="Effective pipeline diameter when two or more pipelines exist between two locations [inch]",
+        )
+        mh.p_HW_loss = Param(
+            model.s_LLA,
+            model.s_T,
+            default=0,
+            initialize=0,
+            within=NonNegativeReals,
+            mutable=True,
+            units=pyunits.meter,
+            doc="Hazen-Williams Frictional loss, m",
+        )
+        for key in model.s_LLA:
+            mh.p_effective_Pipeline_diameter[key] = mh.p_Initial_Pipeline_Diameter[
+                key
+            ] + value(
+                sum(
+                    model.vb_y_Pipeline[key, d]
+                    * model.df_parameters["PipelineDiameterValues"][d]
+                    for d in model.s_D
+                )
+            )
+
+        # Compute Hazen-Williams head
+        for t0 in model.s_T:
+            for key in model.s_LLA:
+                if value(model.v_F_Piped[key, t0]) > 0.01:
+                    if value(mh.p_effective_Pipeline_diameter[key]) > 0.1:
+                        mh.p_HW_loss[key, t0] = _hazen_williams_head(
+                            mh.p_HW_material_factor_pipeline,
+                            pyunits.convert(
+                                model.p_lambda_Pipeline[key], to_units=pyunits.meter
+                            ),
+                            pyunits.convert(
+                                mh.p_effective_Pipeline_diameter[key],
+                                to_units=pyunits.meter,
+                            ),
+                            pyunits.convert(
+                                model.v_F_Piped[key, t0],
+                                to_units=pyunits.m**3 / pyunits.s,
+                            ),
+                        )
+
+        def NodePressureRule(b, l1, l2, t1):
+            if value(model.v_F_Piped[l1, l2, t1]) > 0.01:
+                constraint = (
+                    b.v_Pressure[l1, t1] + model.p_Elevation[l1] * mh.p_rhog
+                    == b.v_Pressure[l2, t1]
+                    + model.p_Elevation[l2] * mh.p_rhog
+                    + b.p_HW_loss[l1, l2, t1] * mh.p_rhog
+                    - b.v_PumpHead[l1, l2, t1] * mh.p_rhog
+                    + b.v_ValveHead[l1, l2, t1] * mh.p_rhog
+                )
+                return process_constraint(constraint)
+            else:
+                return Constraint.Skip
+
```

Comment:
I'm curious if this pressure rule can accurately apply to storage tanks, particularly given the dynamic nature of the tank's head level. Additionally, the significant reduction of fluid velocity within the tank seems to be a factor not fully accounted for (by either the Hazen-Williams equation or the valve head loss calculations).

Reply:
Suggestion: Alternatively, we can consider adding a constraint for storage tanks, where we equate tank head to the total of tank water level and tank elevation, to better represent the static nature of water in the tank. Even if we were to assume a constant tank level, incorporating this new constraint can help represent the change in velocity within the tank. This act like forcing a valve head loss prior to tank.

Reply:
Thanks for the comment. The strategic model doesn't really have an explicit storage tank included in it. The head calculation would require detailed geometry of the storage tank / pond (recall we only see capacities). Now, coming to your point of managing pressure. The node pressure can be viewed as the pressure at which water enters the pipeline system so, the required pressure corrections (if any) will be done at the node level. This is similar to assuming a lumped node for tank and a pump/valve. While you suggest is a nice to have fidelity for the model, it is a detail not necessary at the strategic level, at least for now.

Reply:
Thank you for your response. As we have discussed offline, it is beneficial to consider introducing an additional constraint to more accurately represent the storage pressure and the difference in storage inlet and outlet pressure. One potential way is to add:
 Pstorage(t) = H(t) * rho.g
 where dH/dt = 1/A * dQ/dt ,  
(H: the storage tank water level plus the storage location elevation, Q: the net flow to and out of the storage, A: storage cross sectional area).
However, we may want to set this potential improvement aside for the moment and consider revisiting it with a later PR.

Reply:
Thank you for your response. As we have discussed offline, it is beneficial to consider introducing an additional constraint to more accurately represent the storage pressure and the difference in storage inlet and outlet pressure. One potential way is to add:
 Pstorage(t) = H(t) * rho.g
 where dH/dt = 1/A * dQ/dt ,  
(H: the storage tank water level plus the storage location elevation, Q: the net flow to and out of the storage, A: storage cross sectional area).
However, we may want to set this potential improvement aside for the moment and consider revisiting it with a later PR.